### PR TITLE
Zun: Add capsule create method

### PR DESCRIPTION
**Pull request by kevin-zhaoshuai**
Add capsule create method.

For #700

Create:

https://github.com/openstack/zun/blob/master/zun/api/controllers/v1/capsules.py#L124

### CODE REVIEW

**Comment by kevin-zhaoshuai** [comment]
check

**Comment by kevin-zhaoshuai** [comment]
@jtopjian  Hi, may need your help about coverage test.  I find that the coverage test failed but not familiar about it. Could you help about that failed?

**Comment by jrperritt** [comment]
@kevin-zhaoshuai It's alright, don't worry too much about coverage percentage as long as there are unit/acceptance tests.

**Comment by kevin-zhaoshuai** [comment]
@jrperritt Thanks for help! If so, pls help to review this patch if you have time:-)

File: openstack/container/v1/capsules/template.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package capsules
+
+// Template is a structure that represents OpenStack Heat templates
```
[the remark this answers is not in the document]
Comment (kevin-zhaoshuai) [reply openstack/container/v1/capsules/template.go]: Done

File: openstack/container/v1/capsules/template.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+type Template struct {
+	TE
```
[the remark this answers is not in the document]
Comment (kevin-zhaoshuai) [reply openstack/container/v1/capsules/template.go]: Done. Thanks a lot

File: openstack/container/v1/capsules/utils.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	if jerr := json.Unmarshal(t.Bin, &t.Parsed); jerr != nil {
+		if yerr := yaml.Unmarshal(t.Bin, &t.Parsed); yerr != nil {
+			return ErrInvalidDataFormat{}
+		}
+	}
+	return t.Validate()
+}
+
+// Validate validates the contents of TE
+func (t *TE) Validate() error {
```
[the remark this answers is not in the document]
Comment (kevin-zhaoshuai) [reply openstack/container/v1/capsules/utils.go]: sure. Done

File: openstack/orchestration/v1/stacks/template.go [hunk 4]
```diff
@@ -39,8 +39,8 @@ func (t *Template) Validate() error {
 }
 
 // GetFileContents recursively parses a template to search for urls. These urls
-// are assumed to point to other templates (known in OpenStack Heat as child
-// templates). The contents of these urls are fetched and stored in the `Files`
+// are assumed to point to other templates.
+// The contents of these urls are fetched and stored in the `Files`
```
[the remark this answers is not in the document]
Comment (kevin-zhaoshuai) [reply openstack/orchestration/v1/stacks/template.go]: Sorry make mistake here. Revert it now

**Comment by kevin-zhaoshuai** [comment]
@jtopjian  It is a little wired that the accepttest is always failed due to the identify service failed.
I do nothing this identity service. [Link](http://logs.openlabtesting.org/logs/58/858/0028ca2f40d6ff48196fdf23693e061cf9d901c2/check/gophercloud-acceptance-test/2c7e34f/job-output.txt.gz)
What do your opinion about that ? Thanks a lot !

**Comment by jtopjian** [comment]
@kevin-zhaoshuai It looks like OpenLab is having problems in general. Additionally, OpenLab isn't configured to run Zun tests so any failures are not due to this PR.

**Comment by hongbin** [comment]
@dims Anything we can help to get OpenLab configured to run Zun tests?

**Comment by dims** [comment]
@hongbin y, come to #askopenlab on freenode. let's talk there

**Comment by dims** [comment]
@kevin-zhaoshuai the last accept test is showing the following problem

```
2018-04-12 09:05:02.715175 | ubuntu-xenial | acceptance/openstack/identity/v3/endpoint_test.go:18: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715256 | ubuntu-xenial | acceptance/openstack/identity/v3/endpoint_test.go:39: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715328 | ubuntu-xenial | acceptance/openstack/identity/v3/endpoint_test.go:54: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715400 | ubuntu-xenial | acceptance/openstack/identity/v3/endpoint_test.go:77: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715472 | ubuntu-xenial | acceptance/openstack/identity/v3/projects_test.go:67: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715544 | ubuntu-xenial | acceptance/openstack/identity/v3/projects_test.go:94: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715616 | ubuntu-xenial | acceptance/openstack/identity/v3/projects_test.go:129: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715689 | ubuntu-xenial | acceptance/openstack/identity/v3/projects_test.go:136: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715760 | ubuntu-xenial | acceptance/openstack/identity/v3/service_test.go:16: Fatalf format %v reads arg #1, but call has only 0 args
2018-04-12 09:05:02.715830 | ubuntu-xenial | acceptance/openstack/identity/v3/token_test.go:17: Fatalf format %v reads arg #1, but call has only 0 args
```

which seems to have been fixed already in https://github.com/gophercloud/gophercloud/commit/d38bdaf7706c833e9f32c6b1cac3de2191266ff0#diff-f7b3f58c536939140321bd47ddacadc9

So may be you just need a rebase to latest master?

**Comment by hongbin** [comment]
@dims Thanks for the information. @kevin-zhaoshuai it looks this is an easy fix.

In the long-term, we would like to have Zun enabled in the gophercloud CI so that we can test the Zun support against a devstack deployment. This PR is the first step: https://github.com/theopenlab/openlab-zuul-jobs/pull/142 . Reviews are welcome

**Comment by kevin-zhaoshuai** [comment]
@dims @hongbin  @jtopjian Thanks for your advice about that! I will rebase it to latest master to fix this issue.

**Comment by hongbin** [comment]
Note: Zun is enabled in devstack now https://github.com/theopenlab/openlab-zuul-jobs/pull/147 . In a dummy PR, it shows the job was running fine with Zun enabled https://github.com/gophercloud/gophercloud/pull/927

**Comment by kevin-zhaoshuai** [comment]
@hongbin Thanks a lot:-)

**Comment by jtopjian** [comment]
Superseded by #954

**Pull request closed without merging**